drivers/overlay/ov_network.go

```diff
@@ -1031,22 +1036,25 @@ func (n *network) releaseVxlanID() ([]uint32, error) {
 		}
 	}
 	var vnis []uint32
+	n.Lock()
 	for _, s := range n.subnets {
 		if n.driver.vxlanIdm != nil {
-			vni := n.vxlanID(s)
-			vnis = append(vnis, vni)
-			n.driver.vxlanIdm.Release(uint64(vni))
+			vnis = append(vnis, s.vni)
 		}
+		s.vni = 0
+	}
+	n.Unlock()
 
-		n.setVxlanID(s, 0)
+	for _, vni := range vnis {
+		n.driver.vxlanIdm.Release(uint64(vni))
```

Comment:
why do we need  extra for loop instead of using previous logic where its all done inside one place. What are we trying to achieve by doing this ?

Reply:
Fair question.  The vxlanIdm.Release() operation can do locking and write to store while other overlay networks are coming and going as well.   So, I was thinking that this prevents holding the network lock through all those other potentially blocking operations.  (i.e. collect the work list using the lock and then go through the process of releasing them)   But since this function only gets called while the driver is holding the driver lock, that may be a superfluous optimization.